consensus/reactor.go

```diff
@@ -258,6 +258,7 @@ func (conR *ConsensusReactor) Receive(chID byte, src p2p.Peer, msgBytes []byte)
 			ps.ApplyProposalPOLMessage(msg)
 		case *BlockPartMessage:
 			ps.SetHasProposalBlockPart(msg.Height, msg.Round, msg.Part.Index)
+			// TODO: Move this to the part where block part is completely evaluated
```

Comment:
This sounds like proper separation of concerns. It will create some work at the reactor as we will send message for every added vote and part but probably that's fine.

Reply:
I don't think we can do this because it will cause Receive to block on the ConsensusState which is something we don't want

Reply:
Here's a proposal:

- Make a new channel between Reactor and State 
- Whenever the state adds a new block part or vote, it sends the information on the channel (ie. peer ID, whether its a vote or block)
- Reactor has a new routine that loops over this channel and calls the RecordX methods and the MarkGood when necessary

Reply:
Yeah, it (communication between state and reactor bidirectional) does not have to be synchronous. 👍 to your proposal

Reply:
I also haven't assumed that communication should be synchronous. Great guys, thank's for input, I will implement it like this.